File: helper-package-js/bot.js
const { ACTION, NEUTRAL_ID } = require('./constants');
const { StdInReader } = require('./std-in-reader');

const { Cell } = require('./cell');
const { Map } = require('./map');
const { Position } = require('./position');

exports.Bot = class Bot {
    constructor(name) {
        this.name = name;
    }

    async run() {
        await this.getInit();
        this.sendInit(this.name);

        let turn = 1;
        while (true) {

            await this.getTurn();

            const actions = this.getActions();

            this.sendTurn(actions);

            turn++;
        }
    }

    async getInit() {
        this.StdInReader = new StdInReader();
        this.id = parseInt(await this.getString());
        const map = await this.getString();

        const pMap = JSON.parse(map);

        this.currMap = new Map(map);

        this.flowerCells = [];
        this.beeCells = [];
        this.hiveCells = [];

        for (let row of this.currMap.map) {
            for (let item of row) {
                if (item.potency > 0) {
                    this.flowerCells.push(item);
                }
            }
        }
    }

    sendInit(name) {
        this.sendString(name);
    }

    async getTurn() {
        const currState = await this.getString();
        this.currMap.updateMap(currState);

        this.queenBee = this.currMap.queenBees.find(bee => bee.botId == this.id);
        this.beeCells.length = 0;
        this.hiveCells.length = 0;
        this.totalBees = 0;

        // Get all important cells.
        for (let row of this.currMap.map) {
            for (let item of row) {
                const bee = item.bee;
                if (bee && bee.botId == this.id) {
                    this.beeCells.push(item);
                    this.totalBees += bee.count;
                }

                if (item.ownerId == this.id) {
                    this.hiveCells.push(item);
                }
            }
        }
    }

    sendTurn(actions) {
        this.sendString(this._serializeActions(actions));
    }

    sendString(send) {
        console.log(send);
    }

    async getString() {
        return await this.StdInReader.getLine();
    }

    _serializeActions(actions) {
        const mappedActions = [];
        for (let action of actions) {
            const mappedAction = {
                type: action.type
            };
            switch (action.type) {
                case ACTION.MOVE:
                    mappedAction.x = action.pos.x;
                    mappedAction.y = action.pos.y;
                    // Fallthrough.
                case ACTION.MOVE_QUEEN:
                    mappedAction.move = action.moveOrSpawnAmount;
                    mappedAction.face = action.face;
                    break;
                case ACTION.SPAWN:
                    mappedAction.x = action.pos.x;
                    mappedAction.y = action.pos.y;
                    mappedAction.amount = action.moveOrSpawnAmount;
                    mappedAction.face = action.face;
                    break;
                case ACTION.CREATE_HIVE:
                    break;
            }
            mappedActions.push(mappedAction);
        }
        return JSON.stringify(mappedActions);
    }

    isBesideHiveOrQueen(pos) {
        return this.isBeside(pos, this.queenBee.pos) || this.isBesideHive(pos);
    }

    isBesideHive(pos) {
        for (let hiveCell of this.hiveCells) {
            if (this.isBeside(pos, hiveCell.pos)) {
                return true;
            }
        }
        return false;
    }

    isBeside(pos, target) {
        return pos.equals(target) ||
            this.getBoundedPos(pos.x, pos.y + 1).equals(target) ||
            this.getBoundedPos(pos.x, pos.y - 1).equals(target) ||
            this.getBoundedPos(pos.x + 1, pos.y + 1).equals(target) ||
            this.getBoundedPos(pos.x + 1, pos.y - 1).equals(target) ||
            this.getBoundedPos(pos.x - 1, pos.y + 1).equals(target) ||
            this.getBoundedPos(pos.x - 1, pos.y - 1).equals(target) ||
            this.getBoundedPos(pos.x + 1, pos.y).equals(target) ||
            this.getBoundedPos(pos.x - 1, pos.y).equals(target);
    }

    isBesideEnemy(pos) {
        return this.hasEnemy(pos) ||
            this.hasEnemy(this.getBoundedPos(pos.x + 1, pos.y)) ||
            this.hasEnemy(this.getBoundedPos(pos.x + 1, pos.y + 1)) ||
            this.hasEnemy(this.getBoundedPos(pos.x + 1, pos.y - 1)) ||
            this.hasEnemy(this.getBoundedPos(pos.x - 1, pos.y)) ||
            this.hasEnemy(this.getBoundedPos(pos.x - 1, pos.y + 1)) ||
            this.hasEnemy(this.getBoundedPos(pos.x - 1, pos.y - 1)) ||
            this.hasEnemy(this.getBoundedPos(pos.x, pos.y + 1)) ||
            this.hasEnemy(this.getBoundedPos(pos.x, pos.y - 1));
    }

    hasEnemy(pos) {
        const cellBotId = this.currMap.map[pos.y][pos.x].getBeeBotId();
        return cellBotId != NEUTRAL_ID && cellBotId != this.id;
    }

    getBoundedPos(x, y) {
        return new Position((this.currMap.width + x) % this.currMap.width, (this.currMap.height + y) % this.currMap.height);
    }

    getMinPath(from, to) {
        return this.currMap.getPath(from, to);
    }
};

File: helper_package_python/bot.py
from sys import stdin, stdout
import json

from helper_package_python.constants import ActionType, NEUTRAL_ID
from helper_package_python.map import Map
from helper_package_python.position import Position

class Bot:
	def __init__(self, name):
		self.name = name

	def run(self):
		self.getInit()

		self.sendInit(self.name)

		turn = 1
		while True:
			self.getTurn()

			actions = self.getActions()

			self.sendTurn(actions)

			turn += 1

	def getInit(self):
		self.id = int(self.getString())
		map = self.getString()

		pMap = json.loads(map)
		self.currMap = Map(map)

		self.flowerCells = []
		self.beeCells = []
		self.hiveCells = []

		for row in self.currMap.map:
			for item in row:
				if item.potency > 0:
					self.flowerCells.append(item)

	def sendInit(self, name):
		self.sendString(name)

	def getTurn(self):
		currState = self.getString()
		self.currMap.updateMap(currState)

		self.queenBee = next(queenBee for queenBee in self.currMap.queenBees if queenBee.botId == self.id)
		self.beeCells.clear()
		self.hiveCells.clear()
		self.totalBees = 0

		# Get all important cells
		for row in self.currMap.map:
			for item in row:
				bee = item.bee
				if bee and bee.botId == self.id:
					self.beeCells.append(item)
					self.totalBees += bee.count

				if item.ownerId == self.id:
					self.hiveCells.append(item)

	def sendTurn(self, actions):
		self.sendString(self.__serializeActions(actions))

	def getString(self):
		return stdin.readline()

	def sendString(self, output):
		stdout.write(str(output) + '\n')
		stdout.flush()

	def __serializeActions(self, actions):
		mappedActions = []
		for action in actions:
			mappedAction = {
				'type': action.type,
			}
			if action.type == ActionType.MOVE or action.type == ActionType.SPAWN:
				mappedAction['x'] = action.pos.x
				mappedAction['y'] = action.pos.y
				if action.type == ActionType.MOVE:
					mappedAction['move'] = action.moveOrSpawnAmount
				elif action.type == ActionType.SPAWN:
					mappedAction['amount'] = action.moveOrSpawnAmount
				mappedAction['face'] = action.face
			elif action.type == ActionType.MOVE_QUEEN:
				mappedAction['move'] = action.moveOrSpawnAmount
				mappedAction['face'] = action.face

			mappedActions.append(mappedAction)

		return json.dumps(mappedActions)

	def isBesideHiveOrQueen(self, pos):
		return self.isBeside(pos, self.queenBee.pos) or self.isBesideHive(pos)

	def isBesideHive(self, pos):
		for hiveCell in self.hiveCells:
			if self.isBeside(pos, hiveCell.pos):
				return True

		return False

	def isBeside(self, pos, target):
		return (pos == target) or \
			self.getBoundedPos(pos.x, pos.y + 1) == target or \
			self.getBoundedPos(pos.x, pos.y - 1) == target or \
			self.getBoundedPos(pos.x + 1, pos.y + 1) == target or \
			self.getBoundedPos(pos.x + 1, pos.y - 1) == target or \
			self.getBoundedPos(pos.x - 1, pos.y + 1) == target or \
			self.getBoundedPos(pos.x - 1, pos.y - 1) == target or \
			self.getBoundedPos(pos.x + 1, pos.y) == target or \
			self.getBoundedPos(pos.x - 1, pos.y) == target

	def isBesideEnemy(self, pos):
		return self.hasEnemy(pos) or \
			self.hasEnemy(self.getBoundedPos(pos.x + 1, pos.y)) or \
			self.hasEnemy(self.getBoundedPos(pos.x + 1, pos.y + 1)) or \
			self.hasEnemy(self.getBoundedPos(pos.x + 1, pos.y - 1)) or \
			self.hasEnemy(self.getBoundedPos(pos.x - 1, pos.y)) or \
			self.hasEnemy(self.getBoundedPos(pos.x - 1, pos.y + 1)) or \
			self.hasEnemy(self.getBoundedPos(pos.x - 1, pos.y - 1)) or \
			self.hasEnemy(self.getBoundedPos(pos.x, pos.y + 1)) or \
			self.hasEnemy(self.getBoundedPos(pos.x, pos.y - 1))

	def hasEnemy(self, pos):
		cellBotId = self.currMap.map[pos.y][pos.x].getBeeBotId()
		return cellBotId != NEUTRAL_ID and cellBotId != self.id

	def getBoundedPos(self, x, y):
		return Position((self.currMap.width + x) % self.currMap.width, (self.currMap.height + y) % self.currMap.height)

	def getMinPath(self, start, end):
		return self.currMap.getPath(start, end)
